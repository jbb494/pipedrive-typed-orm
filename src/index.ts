export type {
  PipedriveOrmClient,
  BaseSchema,
  PropertiesFromSchema,
  ItemType,
} from "./types";

export { createPipedriveOrmClient } from "./client/client";
