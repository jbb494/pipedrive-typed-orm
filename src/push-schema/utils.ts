import { Ok, Err } from "ga-ts";

export const getAllFields = async (endpointPath: string) => {
  try {
    return Ok(
      (await (
        await fetch(
          `https://api.pipedrive.com/v1/${endpointPath}?api_token=${process.env.PIPEDRIVE_KEY}`,
          {
            method: "GET",
            headers: {
              "Content-Type": "application/json",
            },
          }
        )
      ).json()) as any
    );
  } catch (e) {
    return Err(new Error("Error fetching allFields", { cause: e }));
  }
};
