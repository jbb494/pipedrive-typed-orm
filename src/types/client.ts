import { Result } from "ga-ts";
import { PropertiesFromSchema } from "./properties";
import { CustomSchema } from "./schema";
import { DeatilsResponse } from "./pipedrive-entities";

export type PipedriveOrmClient<
  CustomSchemaT extends CustomSchema,
  CompleteSchema = PropertiesFromSchema<CustomSchemaT>
> = {
  [P in Extract<keyof CompleteSchema, string> as `post${Capitalize<P>}`]: (
    p: CompleteSchema[P]
  ) => Promise<Result<any, Error>>;
} & {
  [R in Extract<keyof DeatilsResponse, string> as `get${Capitalize<R>}`]: (
    id: number
  ) => Promise<Result<DeatilsResponse[R], Error>>;
};
