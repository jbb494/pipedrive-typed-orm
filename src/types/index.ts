export * from "./schema";
export * from "./client";
export * from "./properties";
