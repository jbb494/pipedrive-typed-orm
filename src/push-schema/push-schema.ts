import {
  BaseSchema,
  CustomPipelines,
  CustomSchema,
  SchemaFile,
} from "src/types";
import { Ok, Err, Result } from "ga-ts";
import { getAllFields } from "./utils";

const sleepTime = 200;

const overWriteProperty = async (
  schema: BaseSchema[keyof BaseSchema],
  endpointPath: string
): Promise<Result<{ added: number; removed: number }, Error>> => {
  const allFields = await getAllFields(endpointPath);

  if (!allFields.ok) {
    return Err(new Error("tap", { cause: allFields.error }));
  }
  const allCustomFields = allFields.value.data.filter((x: any) => x.edit_flag);
  const pipedriveNames = allCustomFields.map((x: any) => x.name);

  const addedFields = Object.entries(schema).filter(
    ([name, _elem]) => !pipedriveNames.includes(name)
  );

  const removedFields = allCustomFields.filter(
    ({ name }: any) => !Object.keys(schema).includes(name)
  );
  try {
    for (let i = 0; i < addedFields.length; i++) {
      const [name, field] = addedFields[i];

      console.log(`Adding: ${name}`);
      const data = await (
        await fetch(
          `https://api.pipedrive.com/v1/${endpointPath}?api_token=${process.env.PIPEDRIVE_KEY}`,
          {
            method: "POST",
            body: JSON.stringify({
              name: name,
              ...field,
            }),
            headers: {
              "Content-Type": "application/json",
            },
          }
        )
      ).json();
      await new Promise((resolve) =>
        setTimeout(resolve, sleepTime)
      ); /* sleep */
    }
  } catch (e) {
    return Err(new Error("Error pushing added fields", { cause: e }));
  }

  try {
    for (let i = 0; i < removedFields.length; i++) {
      const { name, id } = removedFields[i];
      console.log(`Removing: ${name}`);
      const data = await (
        await fetch(
          `https://api.pipedrive.com/v1/${endpointPath}/${id}?api_token=${process.env.PIPEDRIVE_KEY}`,
          {
            method: "DELETE",
            headers: {
              "Content-Type": "application/json",
            },
          }
        )
      ).json();
      await new Promise((resolve) =>
        setTimeout(resolve, sleepTime)
      ); /* sleep */
    }
  } catch (e) {
    return Err(new Error("Error removing added fields", { cause: e }));
  }

  return Ok({ added: addedFields.length, removed: removedFields.length });
};

const pushPipelineSchema = async (
  pipelineSchema: CustomPipelines
): Promise<
  Result<
    Record<"pipelines" | "stages", { added: number; removed: number }>,
    Error
  >
> => {
  let q = 0;
  /* Post pipeline with name*/
  const postPipeline = async (name: string): Promise<{ id: number }> =>
    (
      (await (
        await fetch(
          `https://api.pipedrive.com/v1/pipelines?api_token=${process.env.PIPEDRIVE_KEY}`,
          {
            method: "POST",
            body: JSON.stringify({
              name: name,
            }),
            headers: {
              "Content-Type": "application/json",
            },
          }
        )
      ).json()) as any
    ).data;

  /* Post stage with name and pipelineId*/
  const postStage = async (name: string, pipelineId: number): Promise<void> =>
    (
      (await (
        await fetch(
          `https://api.pipedrive.com/v1/stages?api_token=${process.env.PIPEDRIVE_KEY}`,
          {
            method: "POST",
            body: JSON.stringify({
              name: name,
              pipeline_id: pipelineId,
            }),
            headers: {
              "Content-Type": "application/json",
            },
          }
        )
      ).json()) as any
    ).data;

  /* Get all  pipelines */
  const getPipelines = async (): Promise<Array<{ id: number; name: string }>> =>
    (
      (await (
        await fetch(
          `https://api.pipedrive.com/v1/pipelines?api_token=${process.env.PIPEDRIVE_KEY}`,
          {
            method: "GET",
            headers: {
              "Content-Type": "application/json",
            },
          }
        )
      ).json()) as any
    ).data;

  /* Get stages from one pipelines */
  const getStages = async (
    pipelineId: number
  ): Promise<Array<{ name: string; id: string }>> =>
    (
      (await (
        await fetch(
          `https://api.pipedrive.com/v1/stages?api_token=${process.env.PIPEDRIVE_KEY}&pipeline_id=${pipelineId}`,
          {
            method: "GET",
            headers: {
              "Content-Type": "application/json",
            },
          }
        )
      ).json()) as any
    ).data;

  /* Remove pipeliine with id */
  const removePipeline = async (id: number): Promise<void> =>
    (
      (await (
        await fetch(
          `https://api.pipedrive.com/v1/pipelines?api_token=${process.env.PIPEDRIVE_KEY}`,
          {
            method: "DELETE",
            body: JSON.stringify({
              id,
            }),
            headers: {
              "Content-Type": "application/json",
            },
          }
        )
      ).json()) as any
    ).data;

  /* Remove stages from pipelineId */
  const removeStages = async (ids: Array<string>): Promise<void> =>
    (
      (await (
        await fetch(
          `https://api.pipedrive.com/v1/stages?api_token=${process.env.PIPEDRIVE_KEY}`,
          {
            method: "DELETE",
            body: JSON.stringify({
              ids,
            }),
            headers: {
              "Content-Type": "application/json",
            },
          }
        )
      ).json()) as any
    ).data;

  const pipelineNamesFromSchema = Object.keys(pipelineSchema);

  const pipelinesFromPipedrive = await getPipelines();

  const pipelineNamesFromPipedrive = pipelinesFromPipedrive.map((p) => p.name);

  const pipelinesToAdd = pipelineNamesFromSchema.filter(
    (p) => !pipelineNamesFromPipedrive.includes(p)
  );
  const pipelinesToRemove = pipelinesFromPipedrive.filter(
    (p) => !pipelineNamesFromSchema.includes(p.name)
  );
  const pipelinesUnchanged = pipelinesFromPipedrive.filter((p) =>
    pipelineNamesFromSchema.includes(p.name)
  );

  let addedPipelines = 0;
  let removedPipelines = 0;
  let addedStages = 0;
  let removedStages = 0;

  for (const pipelineName of pipelinesToAdd) {
    try {
      const pipeline = await postPipeline(pipelineName);
      addedPipelines++;
      console.log(`Added pipeline: ${pipelineName}`);

      const stagesFromSchema = pipelineSchema[pipelineName];
      for (const stage of stagesFromSchema) {
        await postStage(stage.stageName, pipeline.id);
        addedStages++;
        console.log(
          `Added stage: ${stage.stageName} to pipeline: ${pipelineName}`
        );
        await new Promise((resolve) =>
          setTimeout(resolve, sleepTime)
        ); /* sleep */
      }
    } catch (error) {
      return Err(
        new Error(`Error adding pipeline or stages for ${pipelineName}`, {
          cause: error,
        })
      );
    }
  }

  /* Pipeline to be removed */
  for (const pipeline of pipelinesToRemove) {
    try {
      const stagesFromPipedrive = (await getStages(pipeline.id)) || [];
      await new Promise((resolve) =>
        setTimeout(resolve, sleepTime)
      ); /* sleep */

      removedStages += stagesFromPipedrive.length;

      await removeStages(stagesFromPipedrive.map((s) => s.id));
      console.log(
        `Removed stages (${stagesFromPipedrive.join(",")}) from pipeline: ${
          pipeline.name
        }`
      );
      await new Promise((resolve) =>
        setTimeout(resolve, sleepTime)
      ); /* sleep */
      await removePipeline(pipeline.id);
      console.log(`Removed pipeline: ${pipeline.id}`);
      await new Promise((resolve) =>
        setTimeout(resolve, sleepTime)
      ); /* sleep */

      removedPipelines++;
    } catch (error) {
      return Err(
        new Error(`Error removing pipeline or stages for ${pipeline.id}`, {
          cause: error,
        })
      );
    }
  }

  /* Pipelines that were unchanged */
  for (const pipeline of pipelinesUnchanged) {
    try {
      const stagesFromPipedrive = (await getStages(pipeline.id)) || [];
      await new Promise((resolve) =>
        setTimeout(resolve, sleepTime)
      ); /* sleep */

      const stagesFromSchema = pipelineSchema[pipeline.name];

      const stagesToAdd = stagesFromSchema.filter(
        (s) => !stagesFromPipedrive.map((p) => p.name).includes(s.stageName)
      );
      const stagesToRemove = stagesFromPipedrive.filter(
        (p) => !stagesFromSchema.map((s) => s.stageName).includes(p.name)
      );

      for (const stage of stagesToAdd) {
        await postStage(stage.stageName, pipeline.id);
        addedStages++;
        console.log(
          `Added stage: ${stage.stageName} to pipeline: ${pipeline.name}`
        );
        await new Promise((resolve) =>
          setTimeout(resolve, sleepTime)
        ); /* sleep */
      }

      await removeStages(stagesToRemove.map((x) => x.id));
      console.log(
        `Removed stages (${stagesToRemove.join(",")}) from pipeline: ${
          pipeline.name
        }`
      );
      await new Promise((resolve) =>
        setTimeout(resolve, sleepTime)
      ); /* sleep */
      removedStages += stagesToRemove.length;
    } catch (error) {
      return Err(
        new Error(`Error adding or removing stages for ${pipeline.name}`, {
          cause: error,
        })
      );
    }
  }

  return Ok({
    pipelines: { added: addedPipelines, removed: removedPipelines },
    stages: { added: addedStages, removed: removedStages },
  });
};

type DeltaType = { added: number; removed: number };

export async function pushToPipedrive(schemaFile: SchemaFile): Promise<
  Result<
    {
      resultLead: DeltaType;
      resultPerson: DeltaType;
      resultPipeline: { stages: DeltaType; pipelines: DeltaType };
    },
    Error
  >
> {
  const schemaOverwrite = schemaFile.custom_fields;
  const pipelineSchema = schemaFile.custom_pipelines;
  const resultLead = await overWriteProperty(
    schemaOverwrite.lead,
    "dealFields"
  );
  const resultPerson = await overWriteProperty(
    schemaOverwrite.person,
    "personFields"
  );

  if (!resultLead.ok || !resultPerson.ok)
    return Err(
      new Error("Error pushing properties", {
        cause: { resultLead, resultPerson },
      })
    );

  if (!pipelineSchema)
    return Ok({
      resultLead: resultLead.value,
      resultPerson: resultPerson.value,
      resultPipeline: {
        pipelines: {
          added: 0,
          removed: 0,
        },
        stages: {
          added: 0,
          removed: 0,
        },
      },
    });

  const resultPipeline = await pushPipelineSchema(pipelineSchema);

  if (!resultPipeline.ok) {
    return Err(
      new Error("Error pushing pipelines", { cause: resultPipeline.error })
    );
  }
  return Ok({
    resultLead: resultLead.value,
    resultPerson: resultPerson.value,
    resultPipeline: resultPipeline.value,
  });
}
