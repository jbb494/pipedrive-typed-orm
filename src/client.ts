import { Ok } from "ga-ts";
import { PipedriveOrmClient, Schema } from "./types";

export const createPipedriveOrmClient = <CustomSchema extends Schema>(
  customFieldsSchema: CustomSchema
): PipedriveOrmClient<CustomSchema> => {
  return new Proxy({} as PipedriveOrmClient<CustomSchema>, {
    get: function (target, prop, receiver) {
      if (typeof prop === "string" && prop.startsWith("post")) {
        return function (p: any) {
          console.log({ p, prop });
          return Promise.resolve(Ok({}));
        };
      }
      return Reflect.get(target, prop, receiver);
    },
  });
};
