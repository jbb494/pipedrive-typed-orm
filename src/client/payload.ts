import { AxiosInstance } from "axios";
import { CustomSchema, PropertiesFromSchema } from "src/types";

export const payloadPropertyToPayloadForPipedrive = async <
  CustomSchemaT extends CustomSchema,
  C extends PropertiesFromSchema<CustomSchemaT>,
  P extends C[keyof PropertiesFromSchema<CustomSchemaT>]
>(
  payload: P,
  pathEndpoint: "dealFields" | "personFields",
  axiosInstance: AxiosInstance
) => {
  const response = await axiosInstance.get(`/${pathEndpoint}`, {
    method: "GET",
    headers: {
      "Accept-Encoding": "*",
    },
  });

  const allFields = response.data as any;
  const allCustomFields = allFields.data
    .filter((x: any) => x.edit_flag)
    .reduce((acc: any, elem: any) => ({ ...acc, [elem.name]: elem }), {});

  const { custom_fields, ...fields } = payload;

  const groupBy = (x: any[], key: string) =>
    x.reduce((acc, b) => ({ ...acc, [b[key]]: b }), {});

  const customFieldsPipedrive = Object.entries(custom_fields || {}).reduce(
    (acc, [key, elem]: any) => ({
      ...acc,
      [allCustomFields[key].key]:
        allCustomFields[key].field_type === "enum" ||
        allCustomFields[key].field_type === "set"
          ? Array.isArray(elem)
            ? elem.map(
                (e) => groupBy(allCustomFields[key].options, "label")[e]?.id
              )
            : groupBy(allCustomFields[key].options, "label")[elem]?.id
          : elem,
    }),
    {}
  );

  const payloadToPipedrive = {
    ...fields,
    ...customFieldsPipedrive,
  };
  return payloadToPipedrive;
};
