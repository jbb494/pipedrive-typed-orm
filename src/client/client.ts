import axios, { AxiosInstance, AxiosResponse } from "axios";
import { Err, Ok, Result } from "ga-ts";
import { PipedriveOrmClient, Schema } from "../types";
import { payloadPropertyToPayloadForPipedrive } from "./payload";

export type PipedriveOrmClientConfig =
  | {
      axiosInstance: AxiosInstance;
    }
  | {
      apiKey: string | undefined;
    };
export const createPipedriveOrmClient = <CustomSchema extends Schema>(
  config: PipedriveOrmClientConfig
): PipedriveOrmClient<CustomSchema> => {
  let axiosInstance;
  if (!("axiosInstance" in config)) {
    if (!("apiKey" in config)) throw new Error("Api key must be set");

    axiosInstance = axios.create({
      baseURL: "https://api.pipedrive.com/v1",
      headers: {
        "Accept-Encoding": "*",
      },
    });
    axiosInstance.defaults.params = {};
    axiosInstance.defaults.params["api_token"] = config.apiKey;
  } else axiosInstance = config.axiosInstance;

  return {
    postLead: async (payload): Promise<Result<any, Error>> => {
      let newPayload;
      try {
        newPayload = await payloadPropertyToPayloadForPipedrive(
          payload as any,
          "dealFields",
          axiosInstance!
        );
      } catch (e) {
        return Err(
          new Error("Error creating payload for pipedrive", { cause: e })
        );
      }

      let responsePost: AxiosResponse;
      try {
        responsePost = await axiosInstance!.post("leads", newPayload);
      } catch (e) {
        return Err(new Error("Error posting lead to pipedrive", { cause: e }));
      }

      return Ok(responsePost.data);
    },
    postPerson: async (payload) => {
      let newPayload;
      try {
        newPayload = await payloadPropertyToPayloadForPipedrive(
          payload as any,
          "personFields",
          axiosInstance!
        );
      } catch (e) {
        return Err(
          new Error("Error creating payload for pipedrive", { cause: e })
        );
      }

      let responsePost: AxiosResponse;
      try {
        responsePost = await axiosInstance!.post("persons", newPayload);
      } catch (e) {
        return Err(
          new Error("Error posting person to pipedrive", { cause: e })
        );
      }

      return Ok(responsePost.data);
    },
    postDeal: async (payload) => {
      let newPayload;
      try {
        newPayload = await payloadPropertyToPayloadForPipedrive(
          payload as any,
          "dealFields",
          axiosInstance!
        );
      } catch (e) {
        return Err(
          new Error("Error creating payload for pipedrive", { cause: e })
        );
      }

      let responsePost: AxiosResponse;
      try {
        responsePost = await axiosInstance!.post("deals", newPayload);
      } catch (e) {
        return Err(new Error("Error posting deal to pipedrive", { cause: e }));
      }

      return Ok(responsePost.data);
    },
    getDeal: async (id) => {
      let responseGet;
      try {
        responseGet = await axiosInstance!.get(`deals/${id}`);
      } catch (e) {
        return Err(new Error("Error getting deal frompipedrive", { cause: e }));
      }

      return Ok(responseGet.data);
    },
    getLead: () => {
      throw new Error("Not implemented");
    },
    getPerson: () => {
      throw new Error("Not implemented");
    },
  };
};
