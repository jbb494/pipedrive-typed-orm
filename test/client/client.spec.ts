import axios, { AxiosInstance } from "axios";
import {
  afterAll,
  beforeAll,
  describe,
  expect,
  it,
  mock,
  spyOn,
} from "bun:test";
import { createPipedriveOrmClient } from "src";
import { PipedriveOrmClientConfig } from "src/client";
import {
  createAxiosInstanceMock,
  deleteLead,
  deletePerson,
} from "./helper-client.spec";
import { scenarioASchema } from "test/push-schema/schemas/complex-schemas";
import { emptySchema } from "test/push-schema/schemas";
import { pushToPipedrive } from "src/push-schema";

describe("client", () => {
  const instance: AxiosInstance = createAxiosInstanceMock("scenarioASchema");

  const config = {
    axiosInstance: instance,
  } satisfies PipedriveOrmClientConfig;

  const client = createPipedriveOrmClient<typeof scenarioASchema>(config);
  const postSpy = spyOn(instance, "post");

  describe("Client with scenarioA", () => {
    describe("Post", () => {
      it("should post lead no custom fields", async () => {
        const result = await client.postLead({
          title: "Title lead",
          person_id: 1256,
        });

        expect(postSpy.mock.lastCall).toMatchSnapshot();
        expect(result.ok).toBe(true);
      });
      it("should post lead with custom fields", async () => {
        const result = await client.postLead({
          title: "Title lead",
          person_id: 1256,
          custom_fields: {
            carMake: "bmw",
            kmsInterested: ["10000", "20000"],
          },
        });

        expect(postSpy.mock.lastCall).toMatchSnapshot();
        expect(result.ok).toBe(true);
      });
      it("should post deal with custom fields", async () => {
        const result = await client.postDeal({
          title: "Title lead",
          person_id: 1256,
          custom_fields: {
            carMake: "bmw",
            kmsInterested: ["10000", "20000"],
          },
        });

        expect(postSpy.mock.lastCall).toMatchSnapshot();
        expect(result.ok).toBe(true);
      });
      it("should post person with custom fields", async () => {
        const result = await client.postPerson({
          name: "Whatever",
          email: "whatever@whatever.es",
          custom_fields: {
            partnerName: "whatever",
            partnerAge: 2,
          },
        });

        expect(postSpy.mock.lastCall).toMatchSnapshot();
        expect(result.ok).toBe(true);
      });
    });
  });
  describe("Get", () => {
    it("Should get deal", async () => {
      const id = 7;
      const result = await client.getDeal(id);

      expect(result.value.data).toMatchSnapshot();
      expect(result.value.success).toBe(true);
      expect(result.ok).toBe(true);
    });
  });

  describe("Non mocked client (post operations)", () => {
    const axiosInstance = axios.create({
      baseURL: "https://api.pipedrive.com/v1",
      headers: {
        "Accept-Encoding": "*",
      },
    });
    axiosInstance.defaults.params = {};
    axiosInstance.defaults.params["api_token"] = process.env.PIPEDRIVE_KEY;

    describe("empty schema", () => {
      const client = createPipedriveOrmClient<typeof emptySchema>({
        apiKey: process.env.PIPEDRIVE_KEY,
      });

      it("should post lead no custom fields", async () => {
        const resultPerson = await client.postPerson({
          name: "Test Bob",
        });
        expect(resultPerson.ok).toBe(true);

        const resultLead = await client.postLead({
          title: "Title lead",
          person_id: resultPerson.value.data.id,
        });
        expect(resultLead.ok).toBe(true);

        const deleteLeadResult = await deleteLead(
          axiosInstance,
          resultLead.value.data.id
        );
        const deletePersonResult = await deletePerson(
          axiosInstance,
          resultPerson.value.data.id
        );

        expect(deleteLeadResult.ok).toBe(true);
        expect(deletePersonResult.ok).toBe(true);
      });
    });
    describe("Scenario A", () => {
      const client = createPipedriveOrmClient<typeof scenarioASchema>({
        apiKey: process.env.PIPEDRIVE_KEY,
      });

      beforeAll(async () => {
        await pushToPipedrive(scenarioASchema);
      });
      afterAll(async () => {
        await pushToPipedrive(emptySchema);
      });
      it.only("Should post lead with custom fields", async () => {
        const resultPerson = await client.postPerson({
          name: "Test Bob",
          custom_fields: {
            partnerName: "Maria",
          },
        });
        expect(resultPerson.ok).toBe(true);

        const resultLead = await client.postLead({
          title: "Title lead",
          person_id: resultPerson.value.data.id,
        });
        expect(resultLead.ok).toBe(true);

        const deleteLeadResult = await deleteLead(
          axiosInstance,
          resultLead.value.data.id
        );
        const deletePersonResult = await deletePerson(
          axiosInstance,
          resultPerson.value.data.id
        );

        expect(deleteLeadResult.ok).toBe(true);
        expect(deletePersonResult.ok).toBe(true);
      });
    });
  });
});
