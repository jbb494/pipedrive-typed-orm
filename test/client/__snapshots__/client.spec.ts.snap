// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`client Client with scenarioA Post should post lead no custom fields 1`] = `
[
  "leads",
  {
    "person_id": 1256,
    "title": "Title lead",
  },
]
`;

exports[`client Client with scenarioA Post should post lead with custom fields 1`] = `
[
  "leads",
  {
    "cd0aaf8e7c0628b6458b938d6f2f42cf2845adf0": 303,
    "ce655bf24c8842e5b756575329ff45ae6d023869": [
      306,
      308,
    ],
    "person_id": 1256,
    "title": "Title lead",
  },
]
`;

exports[`client Client with scenarioA Post should post deal with custom fields 1`] = `
[
  "deals",
  {
    "cd0aaf8e7c0628b6458b938d6f2f42cf2845adf0": 303,
    "ce655bf24c8842e5b756575329ff45ae6d023869": [
      306,
      308,
    ],
    "label_ids": [
      "a",
      "b",
    ],
    "person_id": 1256,
    "title": "Title lead",
  },
]
`;

exports[`client Client with scenarioA Post should post person with custom fields 1`] = `
[
  "persons",
  {
    "659db227caeff5da95d7f7edc5328f1122be0b2f": "whatever",
    "b1d5f7480391784cd557b850eca0063dafd2ed4f": 2,
    "email": "whatever@whatever.es",
    "name": "Whatever",
  },
]
`;

exports[`client Get Should get deal 1`] = `
{
  "2902d3de6b5f39ce619680a3b08716dbbaa413af": null,
  "71f8a93027d49969651400c3a6cefc6e5b80bedc": null,
  "7770017d04dac71db36788041a38437735e7a83c": null,
  "active": true,
  "activities_count": 0,
  "add_time": "2024-01-12 21:37:46",
  "age": {
    "d": 0,
    "h": 0,
    "i": 0,
    "m": 0,
    "s": 1,
    "total_seconds": 1,
    "y": 0,
  },
  "average_stage_progress": 0,
  "average_time_to_won": {
    "d": 0,
    "h": 0,
    "i": 0,
    "m": 0,
    "s": 0,
    "total_seconds": 0,
    "y": 0,
  },
  "cc_email": "renovarent-sandbox+deal7@pipedrivemail.com",
  "close_time": null,
  "creator_user_id": {
    "active_flag": true,
    "email": "renovarent.web@gmail.com",
    "has_pic": 0,
    "id": 19039577,
    "name": "Renovarent",
    "pic_hash": null,
    "value": 19039577,
  },
  "currency": "USD",
  "deleted": false,
  "done_activities_count": 0,
  "email_messages_count": 0,
  "expected_close_date": null,
  "files_count": 0,
  "first_won_time": null,
  "followers_count": 1,
  "formatted_value": "$0",
  "formatted_weighted_value": "$0",
  "id": 7,
  "label": null,
  "last_activity": null,
  "last_activity_date": null,
  "last_activity_id": null,
  "last_incoming_mail_time": null,
  "last_outgoing_mail_time": null,
  "lost_reason": null,
  "lost_time": null,
  "next_activity": null,
  "next_activity_date": null,
  "next_activity_duration": null,
  "next_activity_id": null,
  "next_activity_note": null,
  "next_activity_subject": null,
  "next_activity_time": null,
  "next_activity_type": null,
  "notes_count": 0,
  "org_hidden": false,
  "org_id": null,
  "org_name": null,
  "owner_name": "Renovarent",
  "participants_count": 1,
  "person_hidden": false,
  "person_id": {
    "active_flag": true,
    "email": [
      {
        "label": "work",
        "primary": true,
        "value": "john.doe@example.com",
      },
    ],
    "name": "John Doe",
    "owner_id": 19039577,
    "phone": [
      {
        "label": "",
        "primary": true,
        "value": "1234567890",
      },
    ],
    "value": 17,
  },
  "person_name": "John Doe",
  "pipeline_id": 1,
  "probability": null,
  "products_count": 0,
  "rotten_time": null,
  "stage_change_time": null,
  "stage_id": 1,
  "stage_order_nr": 0,
  "status": "open",
  "stay_in_pipeline_stages": {
    "order_of_stages": [
      1,
    ],
    "times_in_stages": {
      "1": 0,
    },
  },
  "title": "Deal 1",
  "undone_activities_count": 0,
  "update_time": "2024-01-12 21:37:47",
  "user_id": {
    "active_flag": true,
    "email": "renovarent.web@gmail.com",
    "has_pic": 0,
    "id": 19039577,
    "name": "Renovarent",
    "pic_hash": null,
    "value": 19039577,
  },
  "value": 0,
  "visible_to": "3",
  "weighted_value": 0,
  "weighted_value_currency": "USD",
  "won_time": null,
}
`;
